### PR TITLE
208 query data event

### PR DESCRIPTION
**Pull request by orangemug**
A slightly changed API from the original ticket, but I think it makes sense.

### CODE REVIEW

File: lib/db/index.js [hunk 1]
```diff
@@ -220,7 +225,24 @@ Db.prototype.exec = function (query, options) {
 
   this.server.logger.debug('executing query against db ' + this.name + ': ' + query);
 
-  return this.send('command', data);
+  this.emit("beginQuery", {
```
Comment (phpnode) [inline lib/db/index.js]: why not just emit the `data` directly?

File: lib/db/index.js [hunk 2]
```diff
@@ -220,7 +225,24 @@ Db.prototype.exec = function (query, options) {
...
+  var s = Date.now();
+  return this.send('command', data).then(function(res) {
+    var e = Date.now();
+    this.emit("endQuery", {
+      perf: {
+        query: e-s
+      }
+    });
+    return res;
+  }.bind(this));
```
Comment (phpnode) [inline lib/db/index.js]: bluebird has a `.bind()` method which is a lot faster than the native one:
```js
return this.send('command', data)
.bind(this)
.then(....)
```

File: test/bugs/82-emojis.js [hunk 3]
```diff
@@ -30,7 +30,7 @@ describe("Bug #82: db.query errors when parsing emojis ", function () {
     return this.db.query(query)
     .bind(this)
     .spread(function (result) {
-      result.should.eql(1);
+      result.should.eql('1');
```
Comment (phpnode) [inline test/bugs/82-emojis.js]: if `.eql` is now stricter, we should just use `.equal` when comparing primitives

File: lib/db/index.js (hunk 1 above)
Comment (orangemug) [inline lib/db/index.js]: Because `data` also includes `sessionId`, `database` and `db`, I don't thing this API needs all those

**Comment by phpnode** [comment]
Thanks for this, I'd still like to only do this when `this.listenerCount('beforeQuery') > 0` though

File: lib/db/index.js [hunk 4]
```diff
@@ -220,7 +225,24 @@ Db.prototype.exec = function (query, options) {
...
+    query: data.query,
+    mode: data.mode,
+    fetchPlan: data.fetchPlan,
+    limit: data.limit,
+    params: data.params
+  });
+
+  var s = Date.now();
+  return this.send('command', data).then(function(res) {
+    var e = Date.now();
```
Comment (phpnode) [inline lib/db/index.js]: maybe this should go in `.tap()` rather than `.then()`, it's useful to measure the response time even for errors

**Comment by orangemug** [comment]
Surely emit will do this anyway right?

File: test/bugs/82-emojis.js (hunk 3 above)
Comment (orangemug) [inline test/bugs/82-emojis.js]: Both are strict now, and I think we should use a strict assertion because... well why not?

**Comment by phpnode** [comment]
@orangemug it will do, but you're still doing the work to prepare data to send to an event which might have no listeners. It's not a *lot* of work, but it's something - instantiating 2 `Date` objects means 2 context switches / trips to the operating system, but that's trivial, the slightly bigger issue is that by adding another `.then()` handler this promise will not resolve until yet-another-turn of the event loop, would like to avoid that as this is a relatively hot path.

File: lib/db/index.js (hunk 4 above)
Comment (orangemug) [inline lib/db/index.js]: Hmmm good point maybe we should also pass the query object and success/fail to `endQuery`, thoughts?

**Comment by orangemug** [comment]
@phpnode ah I see, ok I've added a commit to check event listener count

**Comment by phpnode** [comment]
travis failure was transient, thanks for this!

**Pull request merged**